### PR TITLE
Added Caltech10k Web Faces dataset

### PR DESCRIPTION
**Pull request by mandroid6**
Dataset Request #318 
@rsepassi @cyfra @Conchylicultor

### CODE REVIEW

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        return tfds.core.DatasetInfo(
+            builder=self,
+            description='',
+            features=tfds.features.FeaturesDict({
+                "image": 
+                    tfds.features.Image(encoding_format='jpeg'),
+                "landmarks": {name: tf.int64 for name in LANDMARK_HEADINGS}
+            }),
+            urls=['http://www.vision.caltech.edu/Image_Datasets/Caltech_10K_WebFaces/'],
+            citation=""
```
Comment (us) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: The datasets must be have citation. Please add citation.

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+The dataset contains images of people collected from the web by typing common given names into 
+Google Image Search. The coordinates of the eyes, the nose and the center of the mouth for each 
+frontal face are provided in a ground truth file. This information can be used to align and crop 
+the human faces or as a ground truth for a face detection algorithm. The dataset has 10,524 human 
+faces of various resolutions and in different settings, e.g. portrait images, groups of people, etc. 
+Profile faces or very low resolution faces are not labeled.
+""" 
+class Caltech10K_WebFaces(tfds.core.GeneratorBasedBuilder):
+
+    VERSION = tfds.core.Version("0.1.0")
```
Comment (us) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: Indentation should be 2 as in other files.

File: tensorflow_datasets/image/caltech10k_webfaces.py (hunk 1 above)
Comment (mandroid6) [reply tensorflow_datasets/image/caltech10k_webfaces.py]: Added

File: tensorflow_datasets/image/caltech10k_webfaces.py (hunk 2 above)
Comment (mandroid6) [reply tensorflow_datasets/image/caltech10k_webfaces.py]: Done

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  def _info(self):
+    return tfds.core.DatasetInfo(
+      builder=self,
+      description='',
+      features=tfds.features.FeaturesDict({
+        "image": tfds.features.Image(encoding_format='jpeg'),
+        "landmarks": {name: tf.int64 for name in LANDMARK_HEADINGS}
+      }),
+      urls=['http://www.vision.caltech.edu/Image_Datasets/Caltech_10K_WebFaces/'],
+      citation=""
```
Comment (us) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: I think you forgot that

Comment (mandroid6) [reply tensorflow_datasets/image/caltech10k_webfaces.py]: Yes, fixed now

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    extracted_dirs = dl_manager.download_and_extract({
+      "images": IMAGE_DATA,
+      "landmarks": LANDMARK_DATA
+    })
+    return [
+      tfds.core.SplitGenerator(
+        name=tfds.Split.TRAIN,
+        num_shards=10,
+        gen_kwargs={
+          "file_id": 0,
```
Comment (us) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: You didn't use file_id why returning that ?

Comment (mandroid6) [reply tensorflow_datasets/image/caltech10k_webfaces.py]: Forgot to remove it after updating logic

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        name=tfds.Split.TRAIN,
+        num_shards=10,
+        gen_kwargs={
+          "extracted_dirs": extracted_dirs,
+        }),
+      tfds.core.SplitGenerator(
+        name=tfds.Split.VALIDATION,
+        num_shards=4,
+        gen_kwargs={
+          "extracted_dirs": extracted_dirs,
```
Comment (Conchylicultor) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: I don't understand. It seems that you are passing the same gen_kwargs to your validations and test set. So all train/valid/test sets will contains the same examples no ?

Comment (mandroid6) [reply tensorflow_datasets/image/caltech10k_webfaces.py]: @Conchylicultor some incorrect replication of logic on my end. This dataset doesn't have any original splits given, so will return a single ``tfds.Split.TRAIN`` split.

Fixed now

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+Pruning Training Sets for Learning of Object Categories
+Anelia Angelova, Yaser Abu-Mostafa, Pietro Perona
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import tensorflow as tf
```
Comment (cyfra) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: import tensorflow.compat.v2 as tf

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+  def _split_generator(self, dl_manager):
+    extracted_dirs = dl_manager.download_and_extract({
+      "images": IMAGE_DATA,
+      "landmarks": LANDMARK_DATA
+    })
+    return [
+      tfds.core.SplitGenerator(
+        name=tfds.Split.TRAIN,
+            num_shards=5,
```
Comment (cyfra) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: remove - num_shards is no longer needed.

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    landmarks_path = extracted_dirs["landmarks"]
+    files = tf.io.gfile.listdir(filedir)
+    
+
+    landmarks = self._process_caltech10k_config_file(landmarks_path)
+
+    for file_name in sorted(files):
+      path = os.path.join(filedir, file_name)
+
+      yield {
```
Comment (cyfra) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: the S3 is the default now - so you have to yield a tuple: of (file_name and everything else)

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+  def _generate_examples(self, extracted_dirs):
+    filedir = os.path.join(extracted_dirs['images'],'images')
+    landmarks_path = extracted_dirs["landmarks"]
+    files = tf.io.gfile.listdir(filedir)
+    
+
+    landmarks = self._process_caltech10k_config_file(landmarks_path)
+
+    for file_name in sorted(files):
```
Comment (cyfra) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: nit: doesn't have to be sorted.

File: tensorflow_datasets/image/caltech10k_webfaces.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    lines = data_raw.split("\n")
+
+    keys = [x.split()[0] for x in lines[:-1]]
+    values = {}
+    # Go over each line (skip the last one, as it is empty).
+    for line in lines[0:-1]:
+      row_values = line.strip().split()
+      # Each row start with the 'file_name' and then space-separated values.
+      values[row_values[0]] = [v for v in row_values[1:]]
+    return keys, values
```
Comment (cyfra) [inline tensorflow_datasets/image/caltech10k_webfaces.py]: please explain more what are the "keys" and "values" here..

keys should be the same as "LANDMARK_HEADINGS" right ?

File: tensorflow_datasets/image/caltech10k_webfaces_test.py [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+from tensorflow_datasets.image import caltech10k_webfaces
+import tensorflow_datasets.testing as tfds_test
+
+
+class Caltech10K_WebFacesTest(tfds_test.DatasetBuilderTestCase):
+  DATASET_CLASS = caltech10k_webfaces.Caltech10K_WebFaces
+  SPLITS = {  
+    "train": 10,
+  }
+
```
Comment (cyfra) [inline tensorflow_datasets/image/caltech10k_webfaces_test.py]: does the test pass ? the CSV file is not present in the fake_examples dir.
